wagtail/actions/move_page.py

```diff
@@ -59,8 +59,11 @@ def _move_page(self, page, target, parent_after):
             MP_MoveHandler(page, target, self.pos).process()
 
             # Treebeard's move method doesn't actually update the in-memory instance,
-            # so we need to work with a freshly loaded one now
-            new_page = Page.objects.get(id=page.id)
+            # so we need to work with a freshly loaded one now. We must retrieve the page
+            # instance from the current transaction's database connection, as the changes
+            # are not yet committed and any other database connection will still see the
+            # old state.
+            new_page = Page.objects.using(connection.alias).get(id=page.id)
```

Comment:
I'm curious how this fixes the issue. Do you have a custom database router that specifies `db_for_write()` and `db_for_read()`?

Using `django.db.connection` would mean it uses the `default` database. This may not work for the other way around, i.e. having the `default` database be the read-only one, and using a different database for the write. Although since we don't pass `using=` to `transaction.atomic()`, it probably always uses the `default` anyway regardless of the router.

Perhaps a better solution is to get the database alias from the context manager. I haven't verified if this works, but it seems the atomic context manager does have a `using` attribute that stores the database alias. So, something like

```diff
-         with transaction.atomic():
+         with transaction.atomic() as atomic_ctx:
```

and

```suggestion
            new_page = Page.objects.using(atomic_ctx.using).get(id=page.id)
```

Reply:
Or, if the model state changes to the write DB after `.process()`, I think a better way would be to use [`Model._state.db`](https://docs.djangoproject.com/en/5.1/ref/models/instances/#django.db.models.Model._state), so:

```suggestion
            new_page = Page.objects.using(page._state.db).get(id=page.id)
```

The `_state` attribute is documented, whereas the `Atomic` context object is private (as per its docstring).

Reply:
Yes, we have a custom database router. Read queries are done in our replica database while writes are done in the default.

the transaction context manager has no object, so what you propose won't work.

Neither will this option; 
`page._state.db`

This will be the replica (read) database, and we actually need the instance from the transaction.

Maybe we should specifically the the write database by doing;
```python
write_db = router.db_for_write(Page)
new_page = Page.objects.using(write_db).get(id=page.id)
```

As treebeard does this as well to determine which database to use for writing. And the current PR will always return default, which might not be correct. So I can make that change instead.